### PR TITLE
Implementing the g-and-k models (uni- and bi-variate).

### PR DESCRIPTION
**Pull request by perdaug**
#### Summary:

The implementations of the uni- and bi-variate g-and-k models.

#### Intended Effect:

Increase ELFI's example's pool:
- The g-and-k models can be treated as a benchmark to test ABC methods;
- The bivariate implementation provides intuition in using ELFI for multi-dimensional data;
- The bivariate implementation infers 9 parameter values.

#### How to Verify:

Run the models with the default settings.

#### Side Effects:

N/A

#### Successful inference in current Notebook examples:

N/A

#### Documentation:

Numpy docstring

#### Reviewer Suggestions: 

The default settings are based on the settings described in the referred papers.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD3 (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: elfi/examples/bignk.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    Parameters
+    ----------
+    y : array_like
+        The yielded points.
+
+    Returns
+    -------
+    array_like
+    """
+    E1 = ss.scoreatpercentile(y, 12.5, axis=1)
```
Comment (kskyten) [inline elfi/examples/bignk.py]: Most numpy and scipy operations are vectorized. You can simplify this with something like:
```python
arr = np.random.randn(100)
E1, E2, E3, E4, E5 = ss.scoreatpercentile(arr, np.linspace(0, 100, 5))
```

File: elfi/examples/gnk.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+            summary = elfi.Summary(fn_summary, m['GNK'],
+                                   name=fn_summary.__name__)
+            fns_summary_chosen.append(summary)
+
+    elfi.Distance('euclidean', *fns_summary_chosen, name='d')
+
+    return m
+
+
+def ss_order(y):
```
Comment (kskyten) [inline elfi/examples/gnk.py]: It seems like these functions are duplicated in `gnk.py` and `bignk.py`. It's probably better to put them in the same file and import the models in the `examples` package with the `__init__.py` file.

**Comment by perdaug** [comment]
With some minor differences, the implementations of the summary statistic functions are indeed identical for both models. I agree that these could be re-tweaked in such a way so that they were accessible from one file.

However, since these implementations are self-contained and the functions would not be used anywhere else, the conciseness would be achieved by sacrificing the readability.

I see the purpose of these implementations to be the following:
- A benchmark; and
- A guidance for an external user to familiarise with the way we implement ELFI's example models.

With respect to the latter, the user could read the Python files as Jupyter notebooks; that is, clearly understand what is happening at each step without the need to go through several files.

The main point of the reply is the emphasis of the conciseness-readability trade-off.

File: elfi/examples/bignk.py (hunk 1 above)
Comment (vuolleko) [inline elfi/examples/bignk.py]: The docs suggest that `ss.scoreatpercentile` will be deprecated, and one should use the faster `np.percentile`.

File: elfi/examples/bignk.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    [1] Drovandi, Christopher C., and Anthony N. Pettitt. "Likelihood-free
+    Bayesian estimation of multivariate quantile distributions."
+    Computational Statistics & Data Analysis 55.9 (2011): 2541-2556.
+    [2] Allingham, David, R. AR King, and Kerrie L. Mengersen. "Bayesian
+    estimation of quantile distributions."Statistics and Computing 19.2
+    (2009): 189-201.
+
+    Parameters
+    ----------
+    a1 : float
```
Comment (vuolleko) [inline elfi/examples/bignk.py]: If batch_size>1, these are arrays.

File: elfi/examples/bignk.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+    """Returns an initialised bivariate g-and-k model.
+
+    Parameters
+    ----------
+    n_obs : int, optional
+        The number of the observed points.
+    true_params : array_like, optional
+        The parameters defining the model.
+    stats_summary : array_like, optional
+        The chosen summary statistics, expressed as a list of strings.
```
Comment (vuolleko) [inline elfi/examples/bignk.py]: Add a list of options here, otherwise they are not visible without reading code.

**Comment by kskyten** [comment]
Readability is indeed important. Perhaps in this case the uni- and bivariate cases are similar enough that they should be in the same file.

**Comment by perdaug** [comment]
Have noticed that the batching is faulty. Will fix the issue and commit the changes to the PR.

**Pull request merged**